tests/e2e/ctl_v3_member_test.go

```diff
@@ -55,6 +58,81 @@ func TestCtlV3MemberUpdatePeerTLS(t *testing.T) {
 	testCtl(t, memberUpdateTest, withCfg(*e2e.NewConfigPeerTLS()))
 }
 
+func TestCtlV3ConsistentMemberList(t *testing.T) {
+	e2e.BeforeTest(t)
+
+	epc, err := e2e.NewEtcdProcessCluster(context.TODO(), t,
+		e2e.WithClusterSize(1),
+		e2e.WithWaitClusterReadyTimeout(0),
```

Comment:
I think that setting cluster ready timeout to zero will make the issue appear even after switching to v3. Reason is that during bootstrap it is expected that both storages are stale. It's effects are more visible for v2 as it can be even 10`000 entries behind, but v3 store might also be couple of seconds behind after the last committed change. 

I expect if you added additional writes interleaved between member changes you will notice the same problem. As so this test is incorrect.

Reply:
> I think that setting cluster ready timeout to zero will make the issue appear even after switching to v3. Reason is that during bootstrap it is expected that both storages are stale. It's effects are more visible for v2 as it can be even 10`000 entries behind, but v3 store might also be couple of seconds behind after the last committed change.


Suggest to read https://github.com/etcd-io/etcd/pull/16084#discussion_r1307902400 again. v3store is always the source of truth, it will never be stale by itself. When we say a member's data is stale, it means it's data is behind of the leader's data; in that case, it can't serve lineariziable request.

Reply:
>  it will never be stale by itself.
Oh, but it is. It's stale compared to the last WAL write. Imagine a situation where there was a lot of changes (including membership), all were marked as committed and applied to v3 store, however member crashed before flushing the data from disk. So when it starts back, even if it bootstraps from v3 store (much shorter number of entries that need to be replayed), it still needs to replay all the changes since last disk flush to v3store. 

I think it might be hard to reproduce as in ext4, fsync not only flushes the selected file descriptor, but all file descriptors https://lwn.net/Articles/842385/. I haven't look at this, but it's possible that WAL fsync is also fsyncing etcd db file.

Reply:
> all were marked as committed and applied to v3 store, however member crashed before flushing the data from disk

If the etcd crashes during flushing the data to disk, then the data isn't really applied. 

The backend commits the bbolt db periodically (100ms if I remember it correctly), if etcd crashes anytime in between, it will lose some "already applied" (note it's only in **memory**) entries, when etcd restars, it will replay the WAL files. It's a normal use case, and it's by design.  I do not see any issue.

Reply:
> it will replay the WAL files. It's a normal use case, and it's by design. I do not see any issue.

Yes, it's not an issue. But I think it will be picked up by this test making it flaky.

Reply:
Note I intentionally send a serializable MemberList request. (line 122 in the clt_v3_member_test.go). But it's just a one-member cluster, so the client shouldn't get stale data even it's serializable request.  So it's a bug.

Reply:
> The test will pick up the lost "already applied" entries, making it flaky.

Still do not got your point how the test is flaky. Pls let me know if you have any confusion or need any clarification.

Reply:
Ok, please clarify. If backend commits the bbolt db periodically (as you said above), what happens in this test if etcd crashes before it flushes bbolt. To clarify, before the crash etcd memory will include a new learner, after the crash etcd reads bbolt that doesn't include new learner. On start etcd will load bbolt that doesn't include learner, and will replay WAL entries to add the learner back.

In this test (with cluster ready timeout set to 0) does etcd serve the serializable membership information before it has replayed the WAL entries?

Reply:
OK, eventually got your point, and make some sense to me. If etcd crashes during a test, I would say majority of the existing e2e test cases are flaky:).

Anyway, updated, please take a look at the last commit https://github.com/etcd-io/etcd/pull/16658/commits/7590dfe56a4144dd2d87389018a018a9c158898a

Reply:
By crashing I didn't mean that test will flake due to random crash, I mean the goroutine "Starting a goroutine to repeatedly restart etcdserver" on lines 90-109.

Reply:
pls see line 82 ~ 84. The membership change is guaranteed to be persisted.

Reply:
Sure, but there is a difference between persisting WAL and persisting DB. Are you saying that you are expecting a graceful shutdown via SIGTERM (send by Restart) to result in DB being persisted?

Additional note, I'm having problems with reproducing the issue using the provided test. Will need to double check.

Reply:
This test, doesn't reproduce the issue for me. Any idea why it doesn't find the issue on my side?

Reply:
> Are you saying that you are expecting a graceful shutdown via SIGTERM (send by Restart) to result in DB being persisted?

YES



> This test, doesn't reproduce the issue for me. Any idea why it doesn't find the issue on my side?

You need to remove the second commit and the third commit, and then run the test,

```
$ git reset --hard HEAD~2
$ make gofail-enable
$ cd tests/e2e/
$ go test -run TestCtlV3ConsistentMemberList -v
```

Reply:
ok, it was obvious. I didn't run `make gofail-enable`. Not a fun of fact that the test doesn't do anything if we forget `make gofail-enable`. If we are adding requirement for e2e tests to use a binary with failpoints, we should validate this (or skip). Recommend this as a followup.

Reply:
> Not a fun of fact that the test doesn't do anything if we forget `make gofail-enable`. If we are adding requirement for e2e tests to use a binary with failpoints, we should validate this (or skip)

Note that the failpoint is enabled statically instead of dynamically. To be clearer, I do not use `e2e.WithGoFailEnabled(true)`.  It will never be skipped in github workflow. But of course, developer might forget `make gofail-enable` when executing it locally. Added a comment.